datacube/api/core.py

```diff
@@ -90,25 +90,53 @@ def normalise_config(config):
 
         self.index = index
 
-    def list_products(self, show_archived=False, with_pandas=True):
+    def list_products(self, with_pandas=True, dataset_count=False):
         """
-        List products in the datacube
+        List all products in the datacube. This will produce a ``pandas.DataFrame``
+        or list of dicts containing useful information about each product, including:
 
-        :param show_archived: include products that have been archived.
-        :param with_pandas: return the list as a Pandas DataFrame, otherwise as a list of dict.
+            'name'
+            'description'
+            'license'
+            'default_crs'
+            'default_resolution'
+            'dataset_count' (optional)
+
+        :param bool with_pandas:
+            Return the list as a Pandas DataFrame. If False, return a list of dicts.
+
+        :param bool dataset_count:
+            Return a "dataset_count" column containing the number of datasets
+            for each product. This can take several minutes on large datacubes.
+            Defaults to False.
+
+        :return: A table or list of every product in the datacube.
         :rtype: pandas.DataFrame or list(dict)
         """
-        rows = [product.to_dict() for product in self.index.products.get_all()]
+        # Read properties from each datacube product
+        cols = [
+            'name',
+            'description',
+            'license',
+            'default_crs',
+            'default_resolution',
+        ]
+        rows = [[getattr(pr, col, None) for col in cols]
+                for pr in self.index.products.get_all()]
+
+        # Optionally compute dataset count for each product and add to row/cols
+        if dataset_count:
+            counts = [count[1] for count in self.index.datasets.count_by_product()]
```

Comment:
Very tricky implicit assumption that the order of products is the same for `index.products.get_all()` and index.datasets.count_by_product()`!

Reply:
Yes, totally true! Do we know what defines the order returned by `index.products.get_all()` and `index.datasets.count_by_product()`? The relevant bits here seem to indicate it's coming directly out of the database:

https://github.com/opendatacube/datacube-core/blob/develop/datacube/index/_products.py#L372-L373
https://github.com/opendatacube/datacube-core/blob/develop/datacube/index/_datasets.py#L671-L677

I can revise and match by name if we think this assumption may not always hold!

Reply:
Yeah, not sure. Since there does not seem to be any explicit documented guarantee that they will be the same, better be on the safe side and match by name.

Reply:
OK, just pushed a commit that ensures order by sorting both lists by name prior to combining - it adds a bit more code than I would ideally like, so let me know if you see an easier way.

I've tested this on the DE Africa datacube as the `dataset_count` functionality works nicely there - on the DEA datacube it works too slowly to be useful (I assume because there's much more data there, but it's a big difference in run time (milliseconds vs several minutes).